File: army/JavaArmy/src/com/maksklaban/armymax/spellcasters/Necromanser.java
package com.maksklaban.armymax.spellcasters;

import com.maksklaban.armymax.units.*;
import com.maksklaban.armymax.exceptions.*;

public class Necromanser extends Spellcaster {
    public Necromanser(String nickName, int damage, int hitPoints, int manaPoints, String title, boolean isBattle) throws CloneNotSupportedException {
        super(nickName, damage, hitPoints, manaPoints, title, isBattle, true);
    }

    public Necromanser(String nickName) throws CloneNotSupportedException {
        this(nickName, 15, 155, 105, "Necromanser", true);
    }

    public void update(Unit other) throws ArmyException {
        this.addHitPoints(other.getHitPointsLimit() / 5);
    }

    public void attack(Unit enemy) throws ArmyException {
        super.attack(this, enemy);

        enemy.attach(this);
    }
}

File: army/Unit/Necromanser.cpp
#include "Necromanser.h"

Necromanser::Necromanser(std::string nickName, int damage, int hitPoints, int manaPoints, std::string title) : Spellcaster(nickName, damage, hitPoints, manaPoints, title, true, true) {}

Necromanser::~Necromanser() {}

void Necromanser::attack(Unit* enemy) {
    Unit::attack(this, enemy);

    this->addSubject(enemy);
}

void Necromanser::update(Subject* enemy) {
    Observer::update(enemy);

    this->addHitPoints(this->getDamage());
}